### PR TITLE
[AutoDiff] Add '@differentiable(reverse)'.

### PR DESCRIPTION
**Pull request by rxwei**
Compiler:
- Add `Forward` and `Reverse` to `DifferentiabilityKind`.
- Expand `DifferentiabilityMask` in `ExtInfo` to 3 bits so that it now holds all 4 cases of `DifferentiabilityKind`.
- Parse `@differentiable(reverse)` and `@differentiable(_forward)` declaration attributes and type attributes.
- Emit a warning for `@differentiable` without `reverse`.
- Emit an error for `@differentiable(_forward)`.
- Rename `@differentiable(linear)` to `@differentiable(_linear)`.
- Make `@differentiable(reverse)` type lowering go through today's `@differentiable` code path. We will specialize it to reverse-mode in a follow-up patch.

ABI:
- Add `Forward` and `Reverse` to `FunctionMetadataDifferentiabilityKind`.
- Extend `TargetFunctionTypeFlags` by 1 bit to store the highest bit of differentiability kind (linear). Note that there is a 2-bit gap in `DifferentiabilityMask` which is reserved for `AsyncMask` and `ConcurrentMask`; `AsyncMask` is ABI-stable so we cannot change that.

_Differentiation module:
- Replace all occurrences of `@differentiable` with `@differentiable(reverse)`.
- Delete `_transpose(of:)`.

Resolves rdar://69980056.

### CODE REVIEW

**Comment by rxwei** [comment]
TODO: Update all the call sites.

File: lib/Serialization/ModuleFormat.h [hunk 1]
```diff
@@ -1884,7 +1886,7 @@ namespace decls_block {
   using DifferentiableDeclAttrLayout = BCRecordLayout<
     Differentiable_DECL_ATTR,
     BCFixed<1>, // Implicit flag.
-    BCFixed<1>, // Linear flag.
+    DifferentiabilityKindField, // Linear flag.
```
Comment (dan-zheng) [inline lib/Serialization/ModuleFormat.h:1889]:
```suggestion
    DifferentiabilityKindField, // Differentiability kind.
```

**Review by dan-zheng — APPROVED** [review]

File: lib/Serialization/ModuleFormat.h (hunk 1 above)
Comment (rxwei) [reply lib/Serialization/ModuleFormat.h]: Thanks. Done.

**Comment by rxwei** [comment]
@swift-ci please test

**Review by joshnewnham — APPROVED** [review]
LGTM

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by rxwei** [comment]
@swift-ci please smoke test

**Comment by rxwei** [comment]
@swift-ci please test source compatibility

**Comment by rxwei** [comment]
@swift-ci please test and merge

**Comment by rxwei** [comment]
@swift-ci please test macOS

**Comment by rxwei** [comment]
apple/swift-syntax#258
@swift-ci Please clean test

**Pull request merged**